### ISSUE TITLE
FillConsole on the final cell in a line moves the internal cursor

### ISSUE THREAD

**Issue by malxau** [microsoft/terminal#4561]
# Environment

```
Windows build number: 10.0.18363.657
Windows Terminal version (if applicable):  0.8.10261.0
```

# Steps to reproduce

Here's a repro program.  I've tried to make this as simple as I can, but the way the visual cursor location is not the same as the reported cursor location makes this strangely confusing:

```
#include <windows.h>
#include <stdio.h>

int main(int argc, char * argv[])
{
    CONSOLE_SCREEN_BUFFER_INFO StartScreenInfo;
    CONSOLE_SCREEN_BUFFER_INFO PostFillScreenInfo;
    CONSOLE_SCREEN_BUFFER_INFO PostWriteScreenInfo;
    HANDLE ConHandle;
    DWORD NumberWritten;

    ConHandle = GetStdHandle(STD_OUTPUT_HANDLE);
    GetConsoleScreenBufferInfo(ConHandle, &StartScreenInfo);

    // Set the console mode to some consistent value.
    SetConsoleMode(ConHandle, ENABLE_PROCESSED_OUTPUT);

    // Move the cursor to the rightmost cell in the line, and sleep to show
    // that as far as the user is concerned, that's where it is.  If this is
    // any cell except the rightmost one, the following test case will behave
    // correctly.
    StartScreenInfo.dwCursorPosition.X = StartScreenInfo.dwSize.X - 1;
    SetConsoleCursorPosition(ConHandle, StartScreenInfo.dwCursorPosition);
    Sleep(3000);

    // Fill the final cell.  This API should not move the cursor.  In conhost,
    // the cursor does not appear to move, but in WT, it visually moves.
    // Sleep for a bit to show that the cursor location displayed to the user
    // has changed.
    FillConsoleOutputCharacter(ConHandle, 'X', 1, StartScreenInfo.dwCursorPosition, &NumberWritten);
    Sleep(3000);

    // Check the location of the cursor according to the API.  This shows that
    // the cursor has not moved.
    GetConsoleScreenBufferInfo(ConHandle, &PostFillScreenInfo);

    // Write to the current cursor location.  This shows that the visual
    // display is right and the API is wrong - the cursor has moved.  In
    // conhost, this Y should overwrite the previous X.  In WT, the Y goes
    // off into the ether.
    WriteConsole(ConHandle, "Y", 1, &NumberWritten, NULL);
    GetConsoleScreenBufferInfo(ConHandle, &PostWriteScreenInfo);

    // According to the API, the cursor was in the same place the whole time.
    printf("\nCursor before fill at %i,%i; after fill at %i,%i; after write at %i,%i\n",
           StartScreenInfo.dwCursorPosition.X,
           StartScreenInfo.dwCursorPosition.Y,
           PostFillScreenInfo.dwCursorPosition.X,
           PostFillScreenInfo.dwCursorPosition.Y,
           PostWriteScreenInfo.dwCursorPosition.X,
           PostWriteScreenInfo.dwCursorPosition.Y);

    return 0;
}
```

# Expected behavior

With the attached repro program, the cursor should move to the right of the line; X should be displayed, and the cursor should not move; Y should overwrite X; and the program should complete that indicates each of the three cursor locations are identical.  Note that changing the dwCursorPositionValue to be anything except the rightmost character will generate this behavior.

# Actual behavior

With WT, the repro displays X, but the visual cursor moves even though the reported cursor location does not.  The next character written is presumably written to the visual cursor location, meaning that it is not displayed.  The reported cursor locations suggest the cursor has not moved, which is clearly not what is happening internally.

**Comment by DHowett-MSFT** [comment]
So, I just gave this a test on the 0.9 prereleases (you're on the selfhost program, right?) and it looks like it works properly. I suspect it was related to #4403. Thoughts?

**Comment by DHowett-MSFT** [comment]
![image](https://user-images.githubusercontent.com/14316954/74414122-a6dc4080-4df5-11ea-8ecf-4d6980aea3b9.png)

(and in the intermediate phase, the cursor is still on top of the X)

**Comment by malxau** [comment]
Agree this looks fixed in 0.9.  Closing.
